sycl/doc/extensions/experimental/sycl_ext_intel_fpga_task_sequence.asciidoc

```diff
@@ -0,0 +1,530 @@
+  sycl_ext_intel_fpga_kernel_interface_properties]
+extension.
+
+== Status
+
+This is an experimental extension specification, intended to provide early
+access to features and gather community feedback.  Interfaces defined in this
+specification are implemented in {dpcpp}, but they are not finalized and may
+change incompatibly in future versions of {dpcpp} without prior notice.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+== Backend support status
+
+The APIs in this extension may be used only on a device that has
+`aspect::ext_intel_fpga_task_sequence`.  The application must check that the
+device has this aspect before submitting a kernel using any of the APIs in this
+extension.  If the application tries to use one of these APIs on a device
+without this aspect, the implementation throws a synchronous exception with the
+`errc::kernel_not_supported` error code when the kernel is submitted to the
+queue.
+
+This extension is currently implemented in `icpx` only for FPGA devices and
+only when using either the Level Zero backend or the Intel FPGA OpenCL backend.
+
+== Overview
+
+Although SYCL supports task-level parallelism through device kernels, there is
+no specific mechanism to express parallelism at a sub-kernel level or to invoke
+additional kernels without involving the host program. In OpenCL 2.0 and beyond,
+kernel invocations can be initiated without host program interaction through
+the `enqueue_kernel` API. C++11 also introduced `std::async()` for task-level
+parallelism. However, neither of these approaches is a good fit for spatial
+compute platforms, which require a clear mapping of parallel constructs to
+spatial instances in the hardware.
+
+Rather than significantly changing the semantics of existing solutions, this
+extension introduces a sub-kernel task-level parallelism interface to SYCL for
+use in device code. It provides the class
+`sycl::ext::intel::experimental::task_sequence`, which is an
+abstraction of an asynchronous sequence of invocations of a callable that can be
+executed asynchronously from the caller. A single object of this class can map
+naturally to a single hardware instance on spatial targets.
+
+NOTE: In this document, we use `task_sequence` to indicate the proposed
+`sycl::ext::intel::experimental::task_sequence`
+
+=== Examples
+
+Here's an example that shows how to define and call two task sequences.
+
+```c++
+using namespace sycl::ext::intel::experimental;
+
+// Return the sum of elements in 'v', from index 's' to 's+sz'
+int vectorSum(float *v, size_t s, size_t sz) {
+  ...
+}
+
+...
+
+  sycl::queue q{...};
+
+  constexpr int kCount = ...;
+  float* in = sycl::malloc_host<float>(kCount, q);
+  float* out = sycl::malloc_host<float>(1, q);
+  ...
+
+  q.single_task([=] {
+    task_sequence<vectorSum> firstHalf;
+    task_sequence<vectorSum> secondHalf;
+    constexpr int halfCount = kCount/2;
+    firstHalf.async(in, 0, halfCount);
```

Comment:
We'd also have to submit the task function to the free function. Do you know what motivated this change in direction?

Reply:
So C++ has generally been moving towards favouring free functions over member functions in many cases, the main motivators are that it de-couples the class from the function, allowing it to be more extensible since the same function can be overloaded for multiple types (existing and future), and also as it better supports ADL.

We did this in SYCL 2020 for group functions, so that the same function can be called with any "group type", and we're considering a similar approach for functions like 'submit` and `'parallel_for` (https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_enqueue_functions.asciidoc).

The reason I suggest it here is that for an interface like this where we are launching a function from the kernel, the `task_sequence` is effectively a kind of executor or scheduler type, and it could be nice to extend this in the future to support other types.